src/modules/python/pbs_hooks/PBS_power.PY

```diff
@@ -76,16 +80,308 @@ def vnodes_enabled(job):
+
+    #Assign default values to attributes
+    power_ramp_rate_enable = False
+    power_on_off_enable = False
+    node_idle_limit = 1800
+    min_node_down_delay = 1800
+    max_jobs_analyze_limit = 100
+    max_concurrent_nodes = 10
+
+    #Now assgin values read from config file
+    if 'power_on_off_enable' in config:
+        power_on_off_enable = config['power_on_off_enable']
+    if 'power_ramp_rate_enable' in config:
+        power_ramp_rate_enable = config['power_ramp_rate_enable']
+    if 'node_idle_limit' in config:
+        node_idle_limit = int(config['node_idle_limit'])
+    if 'min_node_down_delay' in config:
+        min_node_down_delay = int(config['min_node_down_delay'])
+    if 'max_jobs_analyze_limit' in config:
+        max_jobs_analyze_limit = int(config['max_jobs_analyze_limit'])
+    if 'max_concurrent_nodes' in config:
+        max_concurrent_nodes = int(config['max_concurrent_nodes'])
+
+
 # Accept if event not serviceable.
 this_event = pbs.event()
 if this_event.type not in [pbs.EXECJOB_PROLOGUE, pbs.EXECJOB_EPILOGUE,
                            pbs.EXECJOB_BEGIN, pbs.EXECJOB_END,
-                           pbs.EXECHOST_STARTUP, pbs.EXECHOST_PERIODIC]:
+                           pbs.EXECHOST_STARTUP, pbs.EXECHOST_PERIODIC,
+                           pbs.PERIODIC]:
     pbs.logmsg(pbs.LOG_WARNING,
                "Event not serviceable for power provisioning.")
     this_event.accept()
 
 
+if this_event.type == pbs.PERIODIC:
+    vnlist = this_event.vnode_list
+    resvlist = this_event.resv_list
+    freq = this_event.freq
+    time_now = time.time()
+
+    # Parse the config file for power attributes
+    try:
+        parse_config_file()
+    except Exception, e:
+        this_event.reject(str(e)) 
+
+    if power_ramp_rate_enable == 0 and power_on_off_enable == 0:
+        this_event.accept()
+
+    if power_on_off_enable and power_ramp_rate_enable:
+        # Disable ramp rate if power on/off is enabled as well.
+        power_ramp_rate_enable = 0
+        pbs.logmsg(pbs.LOG_WARNING,
+            "Hook config: power_on_off_enable is over-riding power_ramp_rate_enable")
+
+    qselect_cmd = os.path.join(pbs_exec, 'bin', 'qselect')
+    qstat_cmd = os.path.join(pbs_exec, 'bin', 'qstat')
+    dtnow = datetime.datetime.now().strftime("%m%d%H%M")
+    qselect_cmd += " -tt.gt." + str(dtnow)
+    try:
+        p = Popen(qselect_cmd, shell=True, stdout=PIPE, stderr=PIPE)
+        (o, e) = p.communicate()
+        if p.returncode or not o:
+            job_list = []
+        else:
+            job_list = o.splitlines()
+    except OSError, ValueError:
+        job_list = []
+    # Analyze queued jobs and see if any of the nodes are needed in near future
+    exec_vnodes = {}
+    i = 0
+    pattern = '%a %b %d %H:%M:%S %Y'
+    for jobid in job_list:
+        if not jobid:
+            break
+        if i > max_jobs_analyze_limit:
+            break
+        cmd = qstat_cmd + " -f " + jobid
```

Comment:
Just a suggestion : You can use dsv or json option for qstat it will be easier to parse .

Reply:
Fixed.